### PR TITLE
Speed Typing a Region of Code

### PR DESCRIPTION
**Pull request by LaurenceWarne**
This PR is essentially the same as #24 but with conflicts fixed, I've closed that one since not using a feature branch was too much of a pain :slightly_smiling_face: .  Apologies for the large diff, hopefully if you've already reviewed #24 it shouldn't be much different.

Original text:

----------------------------------
Hi! This PR adds a function `speed-type-code-region` which works essentially the same as `speed-type-region`, but preserves any syntax highlighting (apologies for gif quality):

speed type code region demo

Additional changes:

- Use overlays instead of `add-text-properties`
- Use lexical binding
- Instead of hard coding replay and play next functions, introduce buffer local variables which point to functions to be called when the user chooses replay or go next
- Omit play next button if it does not make sense in context (e.g. when speed typing a buffer)

### CODE REVIEW

**Comment by dakra** [comment]
Thanks.
As this is a bit bigger I probably need a few more days to review it.

File: speed-type.el [hunk 1]
```diff
@@ -466,6 +478,74 @@ to (point-min) and (point-max)"
...
+                               text syntax-table font-lock-df go-next-fn)))
+    (speed-type--setup-code text #'replay-fn go-next-fn syntax-table font-lock-df)))
+
+(defun speed-type--get-replay-fn (go-next-fn)
+  "Return a replay function which will use GO-NEXT-FN after completion."
+  (lambda (text) (speed-type--setup text
+                                    :replay-fn (speed-type--get-replay-fn go-next-fn)
+                                    :go-next-fn go-next-fn)))
+
+;;;###autoload
```
Comment (dakra) [inline speed-type.el:524]: I think you can remove the autoload cookie as this is only used when speed-type is already loaded

File: speed-type.el [hunk 2]
```diff
@@ -466,6 +478,74 @@ to (point-min) and (point-max)"
...
+(defun speed-type-code-tab ()
+  "A command to be mapped to TAB when speed typing code."
+  (interactive)
+  (let ((start (point))
+	(end (re-search-forward "[^\t ]" (line-end-position) t)))
+    (goto-char start)
+    (when end (insert (buffer-substring-no-properties start (1- end))))))
+
+;;;###autoload
```
Comment (dakra) [inline speed-type.el:533]: Remove autoload.. Same as above

File: speed-type.el [hunk 3]
```diff
@@ -466,6 +478,74 @@ to (point-min) and (point-max)"
...
+(defun speed-type-code-ret ()
+  "A command to be mapped to RET when speed typing code."
+  (interactive)
+  (when (= (point) (line-end-position))
+    (newline) (move-beginning-of-line nil) (speed-type-code-tab)))
+		     
+;;;###autoload
+(defun speed-type-code-region (start end)
```
Comment (dakra) [inline speed-type.el:541]: Do you think it makes sense to not add an extra command but add something like
this to the existing `speed-type-region`?
```emacs-lisp
(if (derived-mode-p 'prog-mode)
    ... speed-type-code-region ...
  ... normal speed-type-text-region ...
)
```

I imagine every time you call speed-type-region in a code buffer you probably want your version?!

File: speed-type.el [hunk 4]
```diff
@@ -466,6 +478,74 @@ to (point-min) and (point-max)"
...
+    (goto-char start)
+    (when end (insert (buffer-substring-no-properties start (1- end))))))
+
+;;;###autoload
+(defun speed-type-code-ret ()
+  "A command to be mapped to RET when speed typing code."
+  (interactive)
+  (when (= (point) (line-end-position))
+    (newline) (move-beginning-of-line nil) (speed-type-code-tab)))
+
```
Comment (dakra) [inline speed-type.el:539]: Remove trailing whitespace

**Review by dakra — COMMENTED** [review]
I gave it a quick look but didn't look at the actual functionality yet.
Thanks.

File: speed-type.el (hunk 3 above)
Comment (LaurenceWarne) [reply speed-type.el]: Yeah that makes sense

File: speed-type.el [hunk 5]
```diff
@@ -415,38 +411,54 @@ Replacements are found in `speed-type-replace-strings'."
    speed-type-replace-strings
    :initial-value text))
 
-(defun speed-type--setup (text &optional author title lang n-words)
+(cl-defun speed-type--setup
+    (text &key author title lang n-words replay-fn go-next-fn callback)
```
Comment (LaurenceWarne) [inline speed-type.el:415]: `replay-fn`, `go-next-fn` are used to set buffer local variables which are called when a user selects to replay/go next, replacing conditional chains in `speed-type--play-next` (not much change in `speed-type--replay`).

**Comment by LaurenceWarne** [comment]
> Thanks. As this is a bit bigger I probably need a few more days to review it.

No problem at all, I can self-review to highlight some of the important bits if thats helpful

File: speed-type.el [hunk 6]
```diff
@@ -1,4 +1,4 @@
-;;; speed-type.el --- Practice touch and speed typing
+;;; speed-type.el --- Practice touch and speed typing -*- lexical-binding: t -*-
```
Comment (LaurenceWarne) [inline speed-type.el:1]: Needed since `speed-type--get-replay-fn` returns a closure.

File: speed-type.el [hunk 7]
```diff
@@ -159,7 +159,8 @@ Total errors: %d
 (defvar-local speed-type--author nil)
 (defvar-local speed-type--lang nil)
 (defvar-local speed-type--n-words nil)
-(defvar-local speed-type--opened-on-buffer nil)
```
Comment (LaurenceWarne) [inline speed-type.el:162]: `speed-type--go-next-fn` and `speed-type--replay-fn` mean we don't have to track whether the current buffer was created as a result of a `speed-type-buffer` command.

File: speed-type.el [hunk 8]
```diff
@@ -415,38 +410,54 @@ Replacements are found in `speed-type-replace-strings'."
...
+      (setq speed-type--go-next-fn go-next-fn)
+      (when replay-fn
+        (setq speed-type--replay-fn replay-fn))
+      (insert text)
+      (set-buffer-modified-p nil)
+      (switch-to-buffer buf)
+      (goto-char 0)
+      (add-hook 'after-change-functions 'speed-type--change nil t)
+      (add-hook 'first-change-hook 'speed-type--first-change nil t)
+      (when callback (funcall callback))
```
Comment (LaurenceWarne) [inline speed-type.el:459]: Not sure for the big diff here, I think this (459) is the only line change in this function, used to fontify the buffer in the case we're speed typing code.

**Comment by jasonm23** [comment]
Hopefully merged soon. 👍🏽

**Comment by dakra** [comment]
I very quickly tried it again and have a bug where I'm not sure if it's my setup.
If I select a region in a lisp buffer with spaces in the beginning, then the first char after the opening parenthesis is always marked as red. Can you reproduce that @LaurenceWarne or is that something with my setup?
E.g. mark
```
     (setq speed-type--title title)
     (setq speed-type--lang lang)
```
with the leading whitespace. Call `M-x speed-type-region`
then press `tab` `(` and `s` and then the s is marked as failed.

**Comment by LaurenceWarne** [comment]
> I very quickly tried it again and have a bug where I'm not sure if it's my setup. If I select a region in a lisp buffer with spaces in the beginning, then the first char after the opening parenthesis is always marked as red. Can you reproduce that @LaurenceWarne or is that something with my setup? E.g. mark
> 
> ```
>      (setq speed-type--title title)
>      (setq speed-type--lang lang)
> ```
> 
> with the leading whitespace. Call `M-x speed-type-region` then press `tab` `(` and `s` and then the s is marked as failed.

I can reproduce, think it's because smartparens (or some equivalent) is inserting `)` as a matching parenthesis, disabling smartparens fixes the issue for me at least.  Perhaps we could do something like:

```lisp
(when (fboundp #'smartparens-mode) (smartparens-mode -1))
(when (fboundp #'electric-pair-mode) (electric-pair-mode -1))
```
After buffer setup.

**Comment by jasonm23** [comment]
Potentially parinfer may do the same thing or similar.

**Comment by dakra** [comment]
electric-pair-mode is even already disabled in your setup function https://github.com/laurencewarne/speed-type/blob/bf53756f62bbb5bde080c25cf57666a640a86b35/speed-type.el#L509

Let's add a kludge like you suggested for smartparens, parinfer, paredit and lispy?

**Comment by LaurenceWarne** [comment]
> Potentially parinfer may do the same thing or similar.

Yeah and paredit, and so on.  My new idea is to `(setq-local post-self-insert-hook nil)` in `speed-type--setup`, maybe a bit of a hack but I think its better than finding disabling parinfer, paredit, smartparens, etc, thoughts?

**Comment by LaurenceWarne** [comment]
> electric-pair-mode is even already disabled in your setup function https://github.com/laurencewarne/speed-type/blob/bf53756f62bbb5bde080c25cf57666a640a86b35/speed-type.el#L509
> 
> Let's add a kludge like you suggested for smartparens, parinfer, paredit and lispy?

I've pushed the idea from my previous comment: https://github.com/dakra/speed-type/pull/34/commits/dacc11cf18f626c584610e87fb5cd140bdf07344, looks good locally to me, do you think this is any better? - A lot less loc anyway :slightly_smiling_face:

It's also triggered CI :tada:

**Comment by jasonm23** [comment]
Seems like a better option than disabling every edit mode that might
interfere!  Not only less LOC but also safe from any new edit modes that
might appear.

**Comment by LaurenceWarne** [comment]
Awesome, thanks!

**Pull request merged**